### DISCUSSION TITLE
Necessity of vertically resolved mask versus horizontal mask & set_bottom_index

### DISCUSSION THREAD

**Discussion by platipodium** [fabm-model/fabm#67] [Q&A]
In SCHISM (unstructured grid), we typically use a vertically resolved mask, to 

1. mask entire water column of dry elements 
2. mask sub-ocean floor elements of any element, i.e. all below the index of the lowest water level along the vertical per element `ibottom(1:ne)`

Case (1) is covered by using a horizontal mask, the question is whether we need case (2) when we also have set_bottom_index(ibottom).  So does FABM implicitly mask everything with index < ibottom ?   If that was the case, we could switch over to using _FABM_HORIZONTAL_MASK_.

But even if not, perhaps it makes make sense to still use _FABM_HORIZONTAL_MASK_ and ignore that means that sub-seafloor elements are calculated; might that be more efficient than implementing a vertically resolved mask?

**Comment by jornbr** [comment]
In principle, FABM has enough information to skip all points below bottom if you provide the horizontal mask and the index of bottom points. However, I expect that this is not fully implemented (certainly not yet tested) in the case where the vectorized dimension is depth and masks are being used - SCHISM is the only model doing this. Thus, some tweaks in the FABM codebase are likely needed to support this.

**Reply by platipodium** [reply]
Thanks for explaining this.  I guess we'll just test both options and report back.

**Comment by platipodium** [comment]
I have two questions regarding this

1. Is it possible to make the mask setting a user option in CMake, such that `cmake -D_FABM_HORIZONTAL_MASK_=OFF` has an effect, overwriting the setting in `fabm_driver.h`?

2. It seems that within the host code, I don't have access to the `_FABM_HORIZONTAL_MASK_` macro, is that correct?  I wanted to do something like 
```
#ifndef _FABM_HORIZONTAL_MASK_
call fs%model%set_mask(mask,mask_hz)
#else
call fs%model%set_mask(mask_hz)
#endif
```

**Comment by jornbr** [comment]
In the end, all that `fabm_driver.h` does is define a few preprocessor symbols. You can potentially set these too from the ouside when you compile FABM. For instance, if you have a cmake-based host such as SCHISM, you can use cmake's [`add_compile_definitions`](https://cmake.org/cmake/help/latest/command/add_compile_definitions.html) to set `_FABM_HORIZONTAL_MASK_`, potentiallly based on user-configurable cmake [options](https://cmake.org/cmake/help/latest/command/option.html). This would need to be done before incorporating FABM via `add_subdirectory`. For non-cmake-based hosts, you could achieve the same by including the preprocessor definitions in the `FFLAGS` environment variable before compiling FABM,  If needed, you could also customize your `fabm_driver.h` to define or omit particular symbols, depending on other preprocessor symbols set by the host.

Regarding host access to the FABM preprocessor symbols: if you do an `#include "fabm_driver.h"` in the host file where you need them, you will have access to all symbols.
